recipe-client-addon/lib/NormandyApi.jsm

```diff
@@ -145,20 +149,65 @@ this.NormandyApi = {
 
   /**
    * Fetch an array of available actions from the server.
+   * @param filters
+   * @param filters.enabled {boolean} If true, only returns enabled
+   * recipes. Default true.
+   * @resolves {Array}
+   */
+  async fetchRecipes(filters = {enabled: true}) {
+    return this.fetchSignedObjects('recipe', filters);
+  },
+
+  /**
+   * Fetch an array of available actions from the server.
    * @resolves {Array}
    */
-  async fetchActions() {
-    const actionApiUrl = await this.getApiUrl("action-list");
-    const res = await this.get(actionApiUrl);
-    return res.json();
+  async fetchActions(filters = {}) {
+    return this.fetchSignedObjects('action', filters);
   },
 
   async fetchImplementation(action) {
-    const response = await fetch(action.implementation_url);
-    if (response.ok) {
-      return response.text();
+    const implementationUrl = this.absolutify(action.implementation_url);
+
+
+    // fetch implementation
+    const response = await fetch(implementationUrl);
+    if (!response.ok) {
+      throw new Error(
+        `Failed to fetch action implementation for ${action.name}: ${response.status}`
+      );
+    }
+    const responseText = await response.text();
+
+    // Try to verify integrity of the implementation text. If no
+    // integrity value is found, skip. If the integrity value doesn't
+    // match thte content or uses an unknown algorithm, fail.
+    const integrityRe = new RegExp('/([a-z]+[0-9]+)-([a-zA-Z0-9+-]+)/$');
+    const match = integrityRe.exec(implementationUrl);
+    if (match) {
+      const [, algorithm, expectedHash] = match;
+      if (algorithm !== 'sha384') {
+        throw new Error(
+          `Failed to fetch action implemenation for ${action.name}: ` +
+          `Unexpected integrity algorithm, expected "sha384", got ${algorithm}`
+        );
+      }
+
+      // verify integrity hash
+      const hasher = Cc["@mozilla.org/security/hash;1"].createInstance(Ci.nsICryptoHash);
+      hasher.init(hasher.SHA384);
+      const dataToHash = new TextEncoder().encode(responseText);
+      hasher.update(dataToHash, dataToHash.length);
+      const useBase64 = true;
+      const hash = hasher.finish(useBase64).replace(/\+/g, '-').replace(/\//g, '_');
+      if (hash !== expectedHash) {
+        throw new Error(
+          `Failed to fetch action implementation for ${action.name}: ` +
+          `Integrity hash does not match content. Expected ${expectedHash} got ${hash}.`
+        );
+      }
```

Comment:
Yes. I understand this. It is totally a thing I know how it works.

![](http://i.imgur.com/bNhbHRb.gif)